### PR TITLE
Feature/920: Implemented 2D convolution

### PR DESCRIPTION
**Pull request by shahpratham**
## Description

Implemented 2D convolution with distributed kernel support.

Issue resolved: #920 

skip ci

### CODE REVIEW

File: heat/core/signal.py [hunk 1]
```diff
@@ -146,3 +201,239 @@ def convolve(a: DNDarray, v: DNDarray, mode: str = "full") -> DNDarray:
...
+    a = a.astype(promoted_type)
+    v = v.astype(promoted_type)
+
+    if a.shape[0] < v.shape[0] and a.shape[1] < v.shape[1]:
+        a, v = v, a
+
+    if len(a.shape) != 2 or len(v.shape) != 2:
+        raise ValueError("Only 2-dimensional input DNDarrays are allowed")
+    if a.shape[0] < v.shape[0] or a.shape[1] < v.shape[1]:
+        raise ValueError("Filter size must not be greater than the  signal size")
```
Comment (ClaudiaComito) [inline heat/core/signal.py:290]: Is this still valid?

Comment (shahpratham) [reply heat/core/signal.py:290]: yes, when `a` is of dimension (4, 15) and `v` is of dimension (5, 10)

File: heat/core/tests/test_signal.py [hunk 2]
```diff
@@ -119,3 +119,116 @@ def test_convolve(self):
...
+                b = ht.array(np_b, split=1, dtype=ht.int32)
+                conv = ht.convolve2d(a, b, mode=mode)
+                self.assert_array_equal(conv, sc_conv)
+
+        # test edge cases
+        # non-distributed signal, size-1 kernel
+        signal = ht.arange(0, 16).reshape(4, 4).astype(ht.int)
+        alt_signal = ht.arange(16).reshape(4, 4).astype(ht.int)
+        # please review and then remove: here we require (for whathever reason): ht.float32 instead of ht.int
+        kernel = ht.ones(1).reshape((1, 1)).astype(ht.float32)
```
Comment (mrfh92) [inline heat/core/tests/test_signal.py:232]: on my machine this was where the unittests failed if `ht.float32` is replaced by `ht.int`

**Comment by mrfh92** [comment]
Error seems to be as follows: 
```
            # apply torch convolution operator
>           signal_filtered = fc.conv2d(signal, weight)
E           RuntimeError: GET was unable to find an engine to execute this computation
heat/core/signal.py:464: RuntimeError
```

**Review by ClaudiaComito — CHANGES_REQUESTED** [review]
@shahpratham 
`torch.conv2d` only supports Float32, that could be why the GPU tests are "failing to find a computation engine"

https://pytorch.org/docs/stable/generated/torch.nn.Conv2d.html?highlight=conv2d#torch.nn.Conv2d

Tagging @krajsek as I see he's been active recently.

**Comment by mrfh92** [comment]
For 1,2,3 MPI-processes on GPU I get the error: 
```
======================================================================
ERROR: test_convolve2d (heat.core.tests.test_signal.TestSignal)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "****/heat/heat/core/tests/test_signal.py", line 165, in test_convolve2d
    conv = ht.convolve2d(dis_signal, kernel_odd, mode=mode)
  File "****/heat/heat/core/signal.py", line 501, in convolve2d
    signal_filtered = fc.conv2d(signal, weight)
RuntimeError: CUDNN_BACKEND_OPERATION: cudnnFinalize Failed cudnn_status: CUDNN_STATUS_BAD_PARAM

----------------------------------------------------------------------
```

**Comment by ClaudiaComito** [comment]
There seems to be a new problem in the CI:
```
from . import pypocketfft as pfft
E   ImportError: /opt/conda/lib/python3.10/site-packages/torch/lib/../../../.././libstdc++.so.6: version `GLIBCXX_3.4.30' not found (required by /opt/conda/lib/python3.10/site-packages/scipy/fft/_pocketfft/pypocketfft.cpython-310-x86_64-linux-gnu.so)
```

The problem is discussed here among many other places:
https://stackoverflow.com/questions/72540359/glibcxx-3-4-30-not-found-for-librosa-in-conda-virtual-environment-after-tryin

I'm not sure what the best way to handle this is, within our CI images.

**Comment by ClaudiaComito** [comment]
> > There seems to be a new problem in the CI:
> > ```
> > from . import pypocketfft as pfft
> > E   ImportError: /opt/conda/lib/python3.10/site-packages/torch/lib/../../../.././libstdc++.so.6: version `GLIBCXX_3.4.30' not found (required by /opt/conda/lib/python3.10/site-packages/scipy/fft/_pocketfft/pypocketfft.cpython-310-x86_64-linux-gnu.so)
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > The problem is discussed here among many other places: https://stackoverflow.com/questions/72540359/glibcxx-3-4-30-not-found-for-librosa-in-conda-virtual-environment-after-tryin
> > I'm not sure what the best way to handle this is, within our CI images.
> 
> It's the cuda one only. Conda installs some older packages and outputs some warning there: WARNING conda.models.version:get_matcher(556): Using .* with relational operator is superfluous and deprecated and will be removed in a future version of conda. Your spec was 1.7.1._, but conda is ignoring the ._ and treating it as 1.7.1

I don't understand @mtar . What spec?

**Comment by mtar** [comment]
The issue happens only on the Cuda Runner, likely because of outdated libraries installed by conda. The failure on the amd runner is caused by a device mismatch in ht.convolve[2d] instead.

**Comment by mtar** [comment]
@ClaudiaComito
Tests on cuda runner working; now same failures as on amd runner

**Comment by ClaudiaComito** [comment]
I cannot reproduce the CUDA problems locally, but even on CPU I run into different problems with mismatching split axes (and local shapes):

```
======================================================================
======================================================================
ERROR: test_convolve2d (heat.core.tests.test_signal.TestSignal)
ERROR: test_convolve2d (heat.core.tests.test_signal.TestSignal)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/c.comito/devel/local/heat/heat/core/tests/test_signal.py", line 226, in test_convolve2d
    conv = ht.convolve2d(signal, dis_kernel_odd, mode=mode)
  File "/Users/c.comito/devel/local/heat/heat/core/signal.py", line 635, in convolve2d
    signal_filtered += global_signal_filtered[start_idx : start_idx + gshape[0]]
  File "/Users/c.comito/devel/local/heat/heat/core/arithmetics.py", line 188, in add_
    return _operations.__binary_op(wrap_add_, t1, t2, out=t1)
  File "/Users/c.comito/devel/local/heat/heat/core/_operations.py", line 197, in __binary_op
    sanitation.sanitize_out(out, output_shape, output_split, output_device, output_comm)
  File "/Users/c.comito/devel/local/heat/heat/core/sanitation.py", line 300, in sanitize_out
    raise ValueError(f"Expecting output buffer of split {output_split}, got {out.split}")
ValueError: Expecting output buffer of split 0, got None
```

(tests run on 2 processes).

**Comment by ClaudiaComito** [comment]
I'm setting this PR to draft and marking it for next Monday's discussion. My thoughts so far (fed from discussions with @mrfh92 in the past):

- code still needs significant debugging
- fully distributed (signal and kernel) scaling behaviour unclear
- fully distributed via FFT is being implemented (#1291 )

Options for this PR:
1. fix bugs and merge independent of scaling behaviour (needs work)
2. throw out distributed-kernel implementation, keep distributed-signal option only (needs work but will cover most use cases)
3. do not merge
4. ...??


What do you think @mrfh92 @krajsek @mtar ?

**Comment by JuanPedroGHM** [comment]
Work will continue in #1937, closing this one.
